security/pkg/nodeagent/caclient/providers/citadel/client_test.go

```diff
@@ -107,3 +115,108 @@ func TestCitadelClient(t *testing.T) {
 		}
 	}
 }
+
+type mockTokenCAServer struct {
+	Certs []string
+}
+
+func (ca *mockTokenCAServer) CreateCertificate(ctx context.Context, in *pb.IstioCertificateRequest) (*pb.IstioCertificateResponse, error) {
+	targetJWT, err := extractBearerToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if targetJWT != validToken {
+		return nil, fmt.Errorf("token is not valid")
+	}
+	return &pb.IstioCertificateResponse{CertChain: ca.Certs}, nil
+}
+
+func extractBearerToken(ctx context.Context) (string, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", fmt.Errorf("no metadata is attached")
+	}
+
+	authHeader, exists := md[authorizationMeta]
+	if !exists {
+		return "", fmt.Errorf("no HTTP authorization header exists")
+	}
+
+	for _, value := range authHeader {
+		if strings.HasPrefix(value, bearerTokenPrefix) {
+			return strings.TrimPrefix(value, bearerTokenPrefix), nil
+		}
+	}
+
+	return "", fmt.Errorf("no bearer token exists in HTTP authorization header")
+}
+
+// this test is to test whether the server side receive the correct token when
+// we build the CSR sign request
+func TestCitadelClientWithDifferentTypeToken(t *testing.T) {
+	testCases := map[string]struct {
+		server       mockTokenCAServer
+		expectedCert []string
+		expectedErr  string
+		token        string
+	}{
+		"Valid Token": {
+			server:       mockTokenCAServer{Certs: fakeCert},
+			expectedCert: fakeCert,
+			expectedErr:  "",
+			token:        validToken,
+		},
+		"Empty Token": {
+			server:       mockTokenCAServer{Certs: nil},
+			expectedCert: nil,
+			expectedErr:  "rpc error: code = Unknown desc = no HTTP authorization header exists",
+			token:        "",
+		},
+		"InValid Token": {
+			server:       mockTokenCAServer{Certs: []string{}},
+			expectedCert: nil,
+			expectedErr:  "rpc error: code = Unknown desc = token is not valid",
+			token:        fakeToken,
+		},
+	}
+
+	for id, tc := range testCases {
+		t.Run(id, func(t *testing.T) {
+			s := grpc.NewServer()
+			defer s.Stop()
+			lis, err := net.Listen("tcp", mockServerAddress)
+			if err != nil {
+				t.Fatalf("test case [%s]: failed to listen: %v", id, err)
```

Comment:
you don't need to print the "test case [%s]" again as it's already in the t.Run(id, ...).

Reply:
- Does it use token in the very first request to get the certificate and then in the following request switch to use certificate? (or we never use token and the cert is added manually out-of-bound?)

yes. if user not provided with token, after the first we use token, and then  we never use token if the cert exists.